Column resizing

davidpoling opened the pull request:
I had deleted the previous branch I used because of issues with multiple contributors and difficulty with rebasing my commits to have the correct emails. So, I created this branch, 'column-resizing', and cherry-picked the commits I wanted from the 'column-resize' branch.

Apologies for that mess.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-grid/430)
<!-- Reviewable:end -->

Connected to #456

manolo commented:
1. There have been some changes in master,  please merge. You need to remove vaadin-grid.min.js after merging and run gulp gwt again
2.  When adding columns dynamically, it seems that the feature is enabled by default, take a look to the example, here you have a screenshot:  http://g.recordit.co/mFENkon82m.gif 

---

Review status: 0 of 8 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[vaadin-grid.html, line 397 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-KPvVkUI-c3wa5c734oS:-KPvVkUI-c3wa5c734oT:b-lknvfw) ([raw file](https://github.com/vaadin/vaadin-grid/blob/92c258b5b1d0470819269000ad34aef5c669637d/vaadin-grid.html#L397)):*
> ```HTML
>     font-weight: 500;
>     text-align: left;
>     position: relative;
> ```

This introduces a visual issue that you can see in the Column reordering example. Here you have a screenshot, notice the shadow under the first data cell : http://g.recordit.co/NaOcm3BWDW.gif 

You can fix that in css adding:

```
  :host ::content .header-drag-table .vaadin-grid-header,
  .header-drag-table .vaadin-grid-header {
     box-shadow: none;
  }
```

---

*[vaadin-grid.html, line 711 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-KPvWar5JbsYt4vxprjh:-KPvWar674-TkEJh8PrI:b-msm592) ([raw file](https://github.com/vaadin/vaadin-grid/blob/92c258b5b1d0470819269000ad34aef5c669637d/vaadin-grid.html#L711)):*
> ```HTML
>     cursor: ew-resize;
>     height: 100%;
>     width: 2px;
> ```

I feel that 2px is so much. In frozen columns we have one

---

*[demo/columns.html, line 321 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-KPvDEcIj9VhzJAK5jBi:-KPvDEcJxaHXvFv7Wvf9:b-2gribp) ([raw file](https://github.com/vaadin/vaadin-grid/blob/92c258b5b1d0470819269000ad34aef5c669637d/demo/columns.html#L321)):*
> ```HTML
>     <h3>Column Resizing</h3>
>     <p>Columns can be resized. By default, resizing is disabled. You must add the <code>resizable</code> attribute to any desired columns to enable it.
>        In this example, the first column is not resizable.</p>
> ```

Do you might to rephrase this as:

`Columns can be resized by dragging the separation line in headers. To enable resizing, add the <code>resizable</code> attribute to each column you want to resize. In this example, all the columns are resizable but the first one.`

---

*[java/src/main/java/com/vaadin/elements/grid/GridElement.java, line 617 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-KPrHWfhQv3xorUy5Rik:-KPrHWfhQv3xorUy5Ril:b-750vq4) ([raw file](https://github.com/vaadin/vaadin-grid/blob/92c258b5b1d0470819269000ad34aef5c669637d/java/src/main/java/com/vaadin/elements/grid/GridElement.java#L617)):*
> ```Java
>         triggerEvent(COLUMN_ORDER_CHANGED_EVENT);
>     }
>     
> ```

remove spaces

---

*[java/src/main/java/com/vaadin/elements/grid/GridElement.java, line 621 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-KPrHYL-fr6NRGpJGQIr:-KPrHYL-fr6NRGpJGQIs:bkoi2tw) ([raw file](https://github.com/vaadin/vaadin-grid/blob/92c258b5b1d0470819269000ad34aef5c669637d/java/src/main/java/com/vaadin/elements/grid/GridElement.java#L621)):*
> ```Java
>     @Override
>     public void onColumnResize(ColumnResizeEvent<Object> event) {
>     	triggerEvent(COLUMN_RESIZE_CHANGED_EVENT);
> ```

use spaces instead of tabs

---

*[java/src/main/java/com/vaadin/elements/grid/table/GridLightDomTable.java, line 141 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-KPrHf726hFyWgl752GR:-KPrHf726hFyWgl752GS:b-750vq4) ([raw file](https://github.com/vaadin/vaadin-grid/blob/92c258b5b1d0470819269000ad34aef5c669637d/java/src/main/java/com/vaadin/elements/grid/table/GridLightDomTable.java#L141)):*
> ```Java
> 
>             column.setResizable(JSValidate.Boolean.attr($th,  "resizable"));
>             
> ```

remove spaces

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-:-KPvColzdDhL5MgMyjpd:b-wk5570)*
<!-- Sent from Reviewable.io -->

manolo commented:
Also it would be nice if you could document the feature and the new column-resize-changed in asciidoc. Take  as reference `Column Reordering` in `docs/vaadin-grid-columns.adoc`. If you are not very familiar with adocs we will do in a separated commit though. 

---

Review status: 0 of 8 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-:-KPvhN8cwXkclZaJWYZ4:br504fy)*
<!-- Sent from Reviewable.io -->

platosha commented:
The `column-resize-changed` is probably misnamed. By convention in Polymer, the events named as `some-prop-changed` are used to notify of changes for the property `someProp`. In this case, there is no such a property named `columnResize`. I would rename the event to `column-resized`.

Not a blocker for this review, could be also done after merging.

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[vaadin-grid.html, line 711 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-KPvWar5JbsYt4vxprjh:-KPwGk42TUO9rMEl-gBk:b3aw8h0) ([raw file](https://github.com/vaadin/vaadin-grid/blob/92c258b5b1d0470819269000ad34aef5c669637d/vaadin-grid.html#L711)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary>
> I feel that 2px is so much. In frozen columns we have one
</details>
2px looks too thick indeed. But at the same time, 2px is too narrow to conveniently target with mouse, and almost impossible to target on touch devices.

As a fix, we should make a larger touch target area here (my suggestion: at least 8px), while still having a thin nice-looking border.

The cursor should indicate resize on the entire touch target area, not just on the border.

---

*[vaadin-grid.html, line 718 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-KPwIXLCSat_WW1fkzAa:-KPwIXLCSat_WW1fkzAb:bj0er19) ([raw file](https://github.com/vaadin/vaadin-grid/blob/92c258b5b1d0470819269000ad34aef5c669637d/vaadin-grid.html#L718)):*
> ```HTML
>   :host ::content .vaadin-grid-column-resize-handle,
>   .vaadin-grid-column-resize-handle {
>     background-color: var(--primary-color, #e3e3e3);
> ```

It’s not quite correct that we use `--primary-color` with different defaults: the default is grey, and light blue for the active state. When `--primary-color` is defined (e.g., is used with the default theme from `paper-styles`), both colors will be defined as the same, thus making the active state indistinguishable.

I would change this line to use the plain value:
```css
    background-color: #e3e3e3;
```

For the active state, `--primary-color` will be used if defined, but not breaking the default color.

---

*[test/grid-column-resizing.html, line 32 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-KPwNYE_snKM95dSLcYU:-KPwNYE_snKM95dSLcYV:b-q36f3q) ([raw file](https://github.com/vaadin/vaadin-grid/blob/92c258b5b1d0470819269000ad34aef5c669637d/test/grid-column-resizing.html#L32)):*
> ```HTML
>     });
> 
>     it('Mock interactions, resize test (one column)', function () {
> ```

Please name tests in this style:
```javascript
it('should resize a column', function () { /* ... */ }
```

---

*[test/grid-column-resizing.html, line 41 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-KPwOs9-6CuSKjinUr6J:-KPwOs90SlchN0vKFbxQ:b-rfjp2y) ([raw file](https://github.com/vaadin/vaadin-grid/blob/92c258b5b1d0470819269000ad34aef5c669637d/test/grid-column-resizing.html#L41)):*
> ```HTML
> 
>       //get the column resize handle div for the second column
>       var elem = Polymer.dom(grid.root).querySelectorAll('.vaadin-grid-column-resize-handle')[0];
> ```

Small tip: use `querySelector('something')` instead of `querySelectorAll('something')[0]` here and in the test below.

---

*[test/grid-column-resizing.html, line 49 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-KPwNn3Ql7aSZCjJiZG6:-KPwNn3Ql7aSZCjJiZG7:b908qbo) ([raw file](https://github.com/vaadin/vaadin-grid/blob/92c258b5b1d0470819269000ad34aef5c669637d/test/grid-column-resizing.html#L49)):*
> ```HTML
>     });
> 
>     it('Mock interactions, resize test (frozen column)', function () {
> ```

```javascript
it('should resize a frozen column', function () { /* ... */ }
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-:-KPwLSmNlLMjNn6dEOnn:b2xhgwt)*
<!-- Sent from Reviewable.io -->

manolo commented:
Since the other events are misnamed we need a separated PR for fixing them. So @davidpoling  rename yours to column-resized and we will modify the API for the rest.
Also remember to add a comment block about this event in the listeners block so as polymer documents correctly the event:

```
    /**
       * A change in column sizes.
       *
       * @event column-resized
       */
```

---

Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-:-KQ5wJeNEctIcnGBM19i:bmf56v3)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[vaadin-grid.html, line 711 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-KPvWar5JbsYt4vxprjh:-KQ70P_SOm_tNxtXMGEO:b-oarbxz) ([raw file](https://github.com/vaadin/vaadin-grid/blob/92c258b5b1d0470819269000ad34aef5c669637d/vaadin-grid.html#L711)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary>
> 2px looks too thick indeed. But at the same time, 2px is too narrow to conveniently target with mouse, and almost impossible to target on touch devices.
> 
> As a fix, we should make a larger touch target area here (my suggestion: at least 8px), while still having a thin nice-looking border.
> 
> The cursor should indicate resize on the entire touch target area, not just on the border.
</details>
I think we could take the same idea that google spreadsheets. As @platosha  says making the div 8px width, transparent background, and 1px right border. When the mouse hovers, you can change the background and border to blue. Here you have a screenshot: http://g.recordit.co/EB0LXijEHE.gif 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Hi @davidpoling 
What's the status of the PR? Are you still progressing it? If not, we can adopt it and finish it internally.

jstanley1730 commented:
Tomi,

David was pulled onto another project but will be back today.  He is going
to try and finish the PR comments today, if he can't finish up today we
will probably need to turn it over to you guys.  Will you be able to get
this feature into the 1.2 release?

Thanks Jason

On Wed, Aug 31, 2016 at 3:53 AM, Tomi Virkki <notifications@github.com>
wrote:

> Hi @davidpoling <https://github.com/davidpoling>
> What's the status of the PR? Are you still progressing it? If not, we can
> adopt it and finish it internally.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/vaadin/vaadin-grid/pull/430#issuecomment-243686716>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJ0EESbkIUTJUQsFghkhW1wDwN3B6Vjwks5qlTLqgaJpZM4JqDl_>
> .
>



-- 
--Jason Stanley

jouni commented:
We were going to release 1.2 asap with just column reordering, but that is mainly because you need it very soon. It is okay for us to include this as well, and to me it makes more sense to bundle these to together in one minor.

jstanley1730 commented:
Jouni,

Thanks for keeping us in mind, we really appreciate it.  We actually need
both features equally, so some delay to add the column resize would not
hurt us to much.  I am going to David try and clean up the pull request
today, however I don't have much more of his time past this week, so may
need to turn it over to you guys if there are still outstanding issues
after he updates the pull request.

Thanks again

On Wed, Aug 31, 2016 at 10:52 AM, Jouni Koivuviita <notifications@github.com
> wrote:

> We were going to release 1.2 asap with just column reordering, but that is
> mainly because you need it very soon. It is okay for us to include this as
> well, and to me it makes more sense to bundle these to together in one
> minor.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/vaadin/vaadin-grid/pull/430#issuecomment-243789796>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJ0EEdtPzhRF0oVFZz1ixjiqdt84_bFIks5qlZU1gaJpZM4JqDl_>
> .
>



-- 
--Jason Stanley

davidpoling commented:
For some reason when running "gulp gwt," I am getting errors with maven and cannot seem to solve the issue. So, @djbowman1 is building the min.js file for me.

davidpoling commented:
@manolo I don't think I am familiar enough with how the vaadin-grid-column-resize-handle div is created, so I'm not sure if I know how to wrap it in a div for the touch-target ability that you mentioned.  As @jstanley1730 mentioned, I have been moved to another project, but I do believe I was able to address the remaining comments/concerns on this PR as much as I could (added to the .adoc, changed event names, adjusted the width of the resize handle, etc.)

If there is anything else that you would like for me to do, let me know! Thank you!

manolo commented:
@davidpoling Don't worry about those issues. As @tomivirkki said, we can merge the patch and fine tuning after. Thank you very much.

---

Review status: 2 of 9 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-:-KQZDy3ChSbVjx8iZgoO:b-gps170)*
<!-- Sent from Reviewable.io -->

manolo commented:
Reviewed 6 of 6 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430)*
<!-- Sent from Reviewable.io -->

manolo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430#-:-KQZJ0IKFM9dafSH4EVL:bnfp4nl)*
<!-- Sent from Reviewable.io -->

manolo commented:
Reviewed 2 of 8 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/430)*
<!-- Sent from Reviewable.io -->
